sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarPartitionReaderWithPartitionValues.scala

```diff
@@ -95,6 +110,55 @@ object ColumnarPartitionReaderWithPartitionValues {
     result
   }
 
+  /**
+   * Splits each Gpu Column Vector into multiple batches based on row counts.
+   *
+   * Example,
+   * columnVectors = [ Cv('pqr', 55), Cv('111', 55), Cv('ab1', 55)]
+   * rowCounts = [ 20, 10, 15, 10 ]
+   * splitIndices = [ 20, 30, 35 ]
+   * Result:
+   *    [
+   *      [Cv('pqr',20), Cv('pqr',10), Cv('pqr',15), Cv('pqr',10)],
+   *      [Cv('111',20), Cv('111',10), Cv('111',15), Cv('111',10)],
+   *      [Cv('ab1',20), Cv('ab1',10), Cv('ab1',15), Cv('ab1',10)]
+   *    ]
+   */
+  private def splitColumnVector(columnVectors: Seq[ColumnVector],
+      rowCounts: Array[Long]): Seq[Array[GpuColumnVector]] = {
+    // Calculate indices from rowCounts
+    val splitIndices = rowCounts.scanLeft(0) {
+      case (cumulativeCount, partitionIndex) =>
+        // ColumnVector#split() requires indices to be 'int'
+        cumulativeCount + partitionIndex.toInt
+    }.drop(1).dropRight(1)
+
+    columnVectors.map {
+      case gpuColumnVector: GpuColumnVector =>
+        val splitColumns = gpuColumnVector.getBase.split(splitIndices: _*)
+        splitColumns.map(splitCol => GpuColumnVector.from(splitCol, gpuColumnVector.dataType()))
+    }
+  }
+
+  def addGpuColumVectorsToBatchIter(
+     fileBatch: ColumnarBatch,
+     batchOfPartitionColumns: Array[Array[GpuColumnVector]]): Iterator[ColumnarBatch] = {
+    val fileBatchCols = (0 until fileBatch.numCols).map(fileBatch.column)
+    val rowCounts = batchOfPartitionColumns.map(_.head.getRowCount)
+    val columnVectorsSeq = splitColumnVector(fileBatchCols, rowCounts)
```

Comment:
These batch splits need to be made spillable until they are returned by the iterator.

Reply:
Return by iterator will occur in `GpuColumnarBatchWithPartitionValuesIterator#next()` when `addMultiplePartitionValuesAndCloseIter()` is called. 

Should we change the return type of all intermediate methods from `Iterator[ColumnarBatch]` to `Iterator[SplillableColumnarBatch]`?

Reply:
The iterator should be of ColumnarBatch.  The user of the iterator wants a ColumnarBatch.  Internally, the iterator will need to deal with spillable batches in order to avoid leaving GPU memory occupied for things not being returned by the current iteration.  For example, let's say next() is called on the iterator, and it has nothing yet so it in turn calls next on it's parent iterator.  A batch comes in, but it needs to be split into three batches.  We do the split, producing three batches.  The first batch will be sent down, along with the partition columns, as the result of the next() call.  The other two batches need to be turned into spillable columnar batches (and put on a queue or something similar) to be leveraged by the next two next() calls before we pull from the parent iterator again.